### PR TITLE
Added the MUMBO acquisition function

### PR DESCRIPTION
**Pull request by henrymoss**
Hi 

I have added the MUMBO (MUlti-task Max-value Bayesian Optimisation) acquisition function from my recent paper https://arxiv.org/abs/2006.12093.  I used Emukit for all the experiments and was able to substantially outperform the current multi-fidelity entropy-based acquisition function used by Emukit.

This is a multi-fidelity extension of the max-value entropy search (MES) acquisition function and have coded it up in a very similar way to the implementation of MES I provided to Emukit last year.

I have also included a Jupyter notebook, showing how to use MUMBO.

Thanks Henry 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by mmahsereci** [comment]
Thanks @henrymoss , nice to see Mumbo. Some tests seem to be failing, for example `test_acquisition_evaluate_shape`. Can you check the failed tests first please.

**Comment by henrymoss** [comment]
Sorry about that. Passes the tests now. 

There are errors with some of the local penalization tests, but I don't think that is due to me?

**Comment by mmahsereci** [comment]
Yeah, not entirely sure what's going on. There is another error for `Emukit-tutorial-multi-fidelity-MUMBO-Example.ipynb`. It says `E           jupyter_client.kernelspec.NoSuchKernel: No such kernel named emukit`. I could not replicate it so far on my local machine. @apaleyes did you come across this error by any change before?

**Comment by apaleyes** [comment]
Yeah, this one is pretty clear. The notebook has this section:

```
  "kernelspec": {
   "display_name": "emukit",
   "language": "python",
   "name": "emukit"
  },
```

@henrymoss I assume you've created an `emukit` kernel for this notebook? Can you change it to just standard Python 3? It should look something like this:

```
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
```

That would solve the notebook issue

**Comment by apaleyes** [comment]
Thanks for the fix, @henrymoss. The LP tests failure isn't due to you indeed, but this (or any other PR) cannot be merged until this is fixed. See #316 (and #315 too) for more details

**Comment by apaleyes** [comment]
Hey Henry! Apologies for getting to the actula subtance of this PR only now, we had to clear a few hurdles.

We have two options for this code:
1. Either Mumbo class gets significant refactoring (including variable names, functions, etc) and tests. Then it can stay in the core lib.
2. Or it is moved to examples, along with the notebook, where we are far more relaxed and none of that is really required

Your decision!

**Comment by henrymoss** [comment]
I'm happy with the notebook. Although it might be of advantage to Emukit to have a light-weight multi-fidelity acquisition function? 

What exactly would refactoring require?

**Comment by apaleyes** [comment]
@henrymoss you are quite right, it would be nice to have MUMBO in the core lib.

So the code needs either more telling names, or more comments explaining what is going on. You have actually already done decent work on this, but I can drop individual comments on the PR in places where i think more of this is needed.

Also, we would need at least some super basic unit tests. Tests could just very literally put in zeros and get zeros (or whatever) out. Any edge case input would do as long as we can verify the output deterministically. The whole point is to make sure something makes the code run.

**Comment by henrymoss** [comment]
Okay. I will spend time later this week getting it to scratch!

Thanks for the offer of help.

**Comment by henrymoss** [comment]
I'm adding some tests. But please can you add some pointers for improving the interpretability of the PR?

**Comment by henrymoss** [comment]
I will also add some tests for standard MES. It doesn't seem like you currently have any tests for the multi-fidelity version of entropy search?

File: emukit/bayesian_optimization/acquisitions/max_value_entropy_search.py [hunk 1]
```diff
@@ -109,3 +111,184 @@ def evaluate(self, x: np.ndarray) -> np.ndarray:
...
+
+def approx_int(corr, gamma, upper_lim, lower_lim):
+    # helper function to numerically approx differential entropy of ESG in MUMBO
+    # if corr is 1 then can just calc exactly (like MES)
+    if corr == 1:
+        term1 = 0.5 * np.log(2 * np.pi * np.e)
+        term2 = np.log(1 - norm.cdf(gamma))
+        term3 = gamma * norm.pdf(gamma) / (2 * (1 - norm.cdf(gamma)))
+        return np.mean(term1 + term2 + term3)
+    else:
```
Comment (apaleyes) [inline emukit/bayesian_optimization/acquisitions/max_value_entropy_search.py:267]: Would be nice to have some more details about what's going on in these if/else branches

File: emukit/bayesian_optimization/acquisitions/max_value_entropy_search.py [hunk 2]
```diff
@@ -109,3 +111,184 @@ def evaluate(self, x: np.ndarray) -> np.ndarray:
...
+        denom = 1 / (1 - norm.cdf(gammas))
+        # need to account for numerical instability
+        denom[denom == np.inf] = 0
+        means = corrs * (norm.pdf(gammas)) * denom
+        Vars = 1 + corrs * (means) * (gammas - norm.pdf(gammas) * denom)
+        Vars[Vars <= 0] = 0
+
+        # get upper limits for numerical integration
+        upper_lim = means + 8 * np.sqrt(Vars)
+        lower_lim = means - 8 * np.sqrt(Vars)
```
Comment (apaleyes) [inline emukit/bayesian_optimization/acquisitions/max_value_entropy_search.py:249]: what's 8 here? why 8? ideally this shoud look like this:

    # comments explains why 8 is chosen
    number_of_std = 8
    upper_lim = means + whatever_8_represents  * np.sqrt(Vars)

**Review by apaleyes — COMMENTED** [review]
A couple of refactoring ideas below

**Comment by apaleyes** [comment]
> I will also add some tests for standard MES. It doesn't seem like you currently have any tests for the multi-fidelity version of entropy search?

Sorry, got busy with another deadline. Posted a few suggestions in the review. Thanks!

**Comment by henrymoss** [comment]
I have substantially cleaned up the code (both MUMBO and MES). I have been able to share code between the two functions. 

Let me know what you think.

I will do some unit tests now :)

**Comment by henrymoss** [comment]
I have added some small tests.

**Review by apaleyes — APPROVED** [review]
Looks great now, thanks!
